pkg/probe/exec/exec.go

```diff
@@ -59,6 +61,10 @@ func (pr execProber) Probe(e exec.Cmd) (probe.Result, string, error) {
 
 	klog.V(4).Infof("Exec probe response: %q", string(data))
 	if err != nil {
+		// Convert timeout errors into failures.
+		if status.Code(err) == codes.DeadlineExceeded {
```

Comment:
this code assumes that timeout will be implemented at grpc layer, not inside the containerD. It introduces tight coupling between with the implementation.

Did you check if `exec.Exited()` was set to `false` in case of timeout?

Reply:
Thanks for the review.
containerd implements the timeout with context timeout and converts the `context.DeadlineExceeded` to grpc `DeadlineExceeded` as response. There is no `ExecSyncResponse` in case of timeout.
https://github.com/containerd/cri/blob/09d6426f33cac217528158ddc6d254ca7d597a7b/pkg/server/container_execsync.go#L171
https://github.com/containerd/containerd/blob/c75180740937d4b2d44b9c1edc1c27b208e66e32/errdefs/grpc.go#L61-L62

I also thought it's too coupled to check grpc error here but didn't find a proper error struct to represent the timeout error.
How about this option:
At rpc client side, convert grpc DeadlineExceeded to `context.DeadlineExceeded` or a new error struct that implements `ExitError`, or the `CodeExitError` if it's acceptable. If `context.DeadlineExceeded`, the `Probe` method needs to convert it to `Failure` result.
https://github.com/kubernetes/kubernetes/blob/4de511a4cfdb8e49e804be18245911bf3270fb2f/pkg/kubelet/cri/remote/remote_runtime.go#L387-L391

Reply:
There was a discussion at sig node meeting lately. And mentioned this comment: https://github.com/kubernetes/kubernetes/issues/92057#issuecomment-644915304 I think it's worth understanding if the timeout was passed correctly to containerd. This may be a better fix.

Reply:
It looks like https://github.com/kubernetes/kubernetes/issues/92057#issuecomment-644915304 was talking about lifecycle event hander, not probe handler?
For probe handler, I think it's passed to container runtime correctly:
https://github.com/kubernetes/kubernetes/blob/e7ca64fbe16d0c4b6c7b36aecde9cd75042b2828/pkg/kubelet/prober/prober.go#L157-L161
https://github.com/kubernetes/kubernetes/blob/e7ca64fbe16d0c4b6c7b36aecde9cd75042b2828/pkg/kubelet/kuberuntime/kuberuntime_container.go#L826-L827
https://github.com/kubernetes/kubernetes/blob/e7ca64fbe16d0c4b6c7b36aecde9cd75042b2828/pkg/kubelet/cri/remote/remote_runtime.go#L382-L387

Besides, the probe failure event in https://github.com/kubernetes/kubernetes/issues/92464#issue-644561356 `Liveness probe errored: rpc error: code = DeadlineExceeded desc = failed to exec in container: timeout 2s exceeded: context deadline exceeded` has the configured timeout `2s`, which was returned by containerd:
https://github.com/containerd/cri/blob/master/pkg/server/container_execsync.go#L171

Reply:
Thank you for all the context. The fact that grpc timeout creeped all the way here confused me. I like your previous comment on converting grpc code back to some non-grpc failure on CRI side. Sounds like a good solution. 

@dchen1107 @yujuhong what do you think?